<a href="https://colab.research.google.com/github/dcreeder89/Food_Sales_Predictions/blob/main/1_Reeder_Food_Sales_Predictions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Food Sales Predictions
- Christina Reeder
- Updated: 17 Nov 2022

In [226]:
import pandas as pd

In [227]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Loading Data

In [228]:
# Save file path as a variable
filename = '/content/drive/MyDrive/Coding Dojo/Projects/Project 1/sales_predictions.csv'
# Save data in file as a new dataframe
sales_df = pd.read_csv(filename)

In [229]:
# Look at summary of DataFrame's columns
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [230]:
# Preview first 5 rows of data
sales_df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

## Data Cleaning

### 1. How many rows and columns?

In [231]:
# Look at the shape of the sales DataFrame to determine number of rows and coumns
sales_df.shape

(8523, 12)

> This DataFrame has 8523 rows, and 12 columns

### 2. What are the datatypes of each variable?

In [232]:
# Examine datatypes of each column
sales_df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

### 3. Are there duplicates? If so, drop any duplicates.

In [233]:
# Check sum of the number of duplicate rows in the dataframe
sales_df.duplicated().sum()

0

> There are no duplicated rows in the dataframe 

### 4. Identify missing values

In [234]:
# Check sum of any data that is NaN
sales_df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

> Item_ Weight has 1463 missing values, and Outlet_Size has 2410 missing values out of the total 8523 rows of data in this DataFrame.

### 5. Decide on how to address missing values and do it.

#### Item_Weight Missing Values

In [235]:
sales_df.sort_values(by='Item_Identifier')

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
8043           DRA12        11.60          Low Fat         0.068535   
118            DRA12        11.60          Low Fat         0.041178   
1693           DRA12        11.60               LF         0.000000   
1197           DRA12        11.60          Low Fat         0.000000   
1245           DRA12        11.60          Low Fat         0.040912   
...              ...          ...              ...              ...   
7926           NCZ54        14.65          Low Fat         0.083489   
214            NCZ54        14.65          Low Fat         0.000000   
1896           NCZ54          NaN          Low Fat         0.082956   
7166           NCZ54          NaN          Low Fat         0.145952   
7738           NCZ54        14.65          Low Fat         0.083699   

        Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
8043  Soft Drinks  143.0154            OUT010                       1998   
118   Soft Drinks  140.3154            OUT017                       2007   
1693  Soft Drinks  141.9154            OUT035                       2004   
1197  Soft Drinks  141.6154            OUT045                       2002   
1245  Soft Drinks  142.3154            OUT013                       1987   
...           ...       ...               ...                        ...   
7926    Household  161.3552            OUT049                       1999   
214     Household  161.5552            OUT010                       1998   
1896    Household  164.0552            OUT027                       1985   
7166    Household  160.9552            OUT019                       1985   
7738    Household  163.4552            OUT018                       2009   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
8043         NaN               Tier 3      Grocery Store           283.6308  
118          NaN               Tier 2  Supermarket Type1          2552.6772  
1693       Small               Tier 2  Supermarket Type1           992.7078  
1197         NaN               Tier 2  Supermarket Type1          3829.0158  
1245        High               Tier 3  Supermarket Type1          2552.6772  
...          ...                  ...                ...                ...  
7926      Medium               Tier 1  Supermarket Type1          7148.0288  
214          NaN               Tier 3      Grocery Store           324.9104  
1896      Medium               Tier 3  Supermarket Type3          5523.4768  
7166       Small               Tier 1      Grocery Store           162.4552  
7738      Medium               Tier 3  Supermarket Type2          2599.2832  

[8523 rows x 12 columns]

> NaNs in the item weight column are randomly missing, but the same items are in different stores, so we can fix the missing values using the known values from other stores.

In [236]:
# Create a series of all the item weights sorted by the item identifiers
item_weights = sales_df.groupby('Item_Identifier')['Item_Weight'].max()

In [237]:
# convert item_weights to a dictionary
item_weights.to_dict();

In [238]:
for key, cost in item_weights.items():
  # Create filter for the key so that we can change the values in the dataframe
  key_filter = sales_df['Item_Identifier'] == key

  # Change the values in the dataframe that are nan with the actual values of the item weights
  sales_df.loc[key_filter, 'Item_Weight'] = cost

In [239]:
# Check number of NaN values in Item_Weight column
sales_df['Item_Weight'].isna().sum()

4

In [240]:
# Check number of NaN values in the item_weights dictionary from above to see if it matches
item_weights.isna().sum()

4

In [241]:
# Determine which items still have no weight by sorting dataframe
sales_df.sort_values(by='Item_Weight').tail()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
2368           FDC02        21.35          Low Fat         0.068809   
927            FDN52          NaN          Regular         0.130933   
1922           FDK57          NaN          Low Fat         0.079904   
4187           FDE52          NaN          Regular         0.029742   
5022           FDQ60          NaN          Regular         0.191501   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
2368        Canned  258.5278            OUT035                       2004   
927   Frozen Foods   86.9198            OUT027                       1985   
1922   Snack Foods  120.0440            OUT027                       1985   
4187         Dairy   88.9514            OUT027                       1985   
5022  Baking Goods  121.2098            OUT019                       1985   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
2368       Small               Tier 2  Supermarket Type1          5206.5560  
927       Medium               Tier 3  Supermarket Type3          1569.9564  
1922      Medium               Tier 3  Supermarket Type3          4434.2280  
4187      Medium               Tier 3  Supermarket Type3          3453.5046  
5022       Small               Tier 1      Grocery Store           120.5098

In [242]:
# Check 4 items with weight = NaN
sales_df[sales_df['Item_Identifier']=='FDN52']

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
927           FDN52          NaN          Regular         0.130933   

        Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
927  Frozen Foods   86.9198            OUT027                       1985   

    Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
927      Medium               Tier 3  Supermarket Type3          1569.9564

In [243]:
sales_df[sales_df['Item_Identifier']=='FDK57']

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
1922           FDK57          NaN          Low Fat         0.079904   

        Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
1922  Snack Foods   120.044            OUT027                       1985   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
1922      Medium               Tier 3  Supermarket Type3           4434.228

In [244]:
sales_df[sales_df['Item_Identifier']=='FDE52']

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility Item_Type  \
4187           FDE52          NaN          Regular         0.029742     Dairy   

      Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
4187   88.9514            OUT027                       1985      Medium   

     Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
4187               Tier 3  Supermarket Type3          3453.5046

In [245]:
sales_df[sales_df['Item_Identifier']=='FDQ60']

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
5022           FDQ60          NaN          Regular         0.191501   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
5022  Baking Goods  121.2098            OUT019                       1985   

     Outlet_Size Outlet_Location_Type    Outlet_Type  Item_Outlet_Sales  
5022       Small               Tier 1  Grocery Store           120.5098

> The four items left in our dataframe with NaN vlaues in the Item_Weight column are the only items with these specific Item_Identifier. Therefore, I am going to drop these four rows from our database since I have no good way to predict the Item Weight for these items.

In [246]:
# Drop rows with NaN in the Item_Weight column
sales_df.dropna(subset=['Item_Weight'], inplace=True)

#### Outlet_Size Missing Values

In [247]:
# Check sum of Outlet_Size strings when grouped by Outlet_Identifier and Outlet_Type
sales_df.groupby(['Outlet_Identifier','Outlet_Type'])['Outlet_Size'].sum()

Outlet_Identifier  Outlet_Type      
OUT010             Grocery Store                                                        0
OUT013             Supermarket Type1    HighHighHighHighHighHighHighHighHighHighHighHi...
OUT017             Supermarket Type1                                                    0
OUT018             Supermarket Type2    MediumMediumMediumMediumMediumMediumMediumMedi...
OUT019             Grocery Store        SmallSmallSmallSmallSmallSmallSmallSmallSmallS...
OUT027             Supermarket Type3    MediumMediumMediumMediumMediumMediumMediumMedi...
OUT035             Supermarket Type1    SmallSmallSmallSmallSmallSmallSmallSmallSmallS...
OUT045             Supermarket Type1                                                    0
OUT046             Supermarket Type1    SmallSmallSmallSmallSmallSmallSmallSmallSmallS...
OUT049             Supermarket Type1    MediumMediumMediumMediumMediumMediumMediumMedi...
Name: Outlet_Size, dtype: object

> OUT010 is Type Grocery Store: the only other outlet with type grocery store is a size 'Small', so we will change the size of all produce in OUT010 to 'Outlet_Size' = 'Small'

> OUT017 and OUT045 are both Supermarket Type1. The other Supermarket Type1 outlets are of the following sizes: Small x 2, Medium x 1, Large x 1. I will look at a couple other datapoints to determine how to proceed with these

##### Supermarket Type1 Deep Dive

In [248]:
# Make a filter to only look at 'Supermarket Type1' Outlets
st1_filter = sales_df['Outlet_Type'] == 'Supermarket Type1'
# Create Supermarket Type1 exclusive dataframe to look at
st1_df = sales_df[st1_filter]

In [249]:
# Look at Outlet_Location_Type and Outlet_Identifier vs. size for all Supermarket Type1 in new dataframe
st1_df.groupby(['Outlet_Location_Type','Outlet_Identifier'])['Outlet_Size'].sum()

Outlet_Location_Type  Outlet_Identifier
Tier 1                OUT046               SmallSmallSmallSmallSmallSmallSmallSmallSmallS...
                      OUT049               MediumMediumMediumMediumMediumMediumMediumMedi...
Tier 2                OUT017                                                               0
                      OUT035               SmallSmallSmallSmallSmallSmallSmallSmallSmallS...
                      OUT045                                                               0
Tier 3                OUT013               HighHighHighHighHighHighHighHighHighHighHighHi...
Name: Outlet_Size, dtype: object

> The only other Tier 2 Supermarket Type1 is size 'Small'. Therefore, I am going to make the educated guess that OUT017 and OUT045 are both 'Small' size, as they are also Tier 2 Location Types

##### Change OUT017 and OUT045 to 'Outlet_Size' of 'Small' in accordance to deep dive above

In [250]:
# Create filter in main dataframe for Outlet_Identifiers of OUT017 and OUT045
OUT017_filter = sales_df['Outlet_Identifier'] == 'OUT017' 
OUT045_filter = sales_df['Outlet_Identifier'] == 'OUT045'

# Change 'Outlet_Size' of all produce in OUT017 and OUT045 to 'Small'
sales_df.loc[OUT017_filter, 'Outlet_Size'] = 'Small'
sales_df.loc[OUT045_filter, 'Outlet_Size'] = 'Small'

# Check that the sizes of the outlets have changed
sales_df.loc[OUT017_filter].head()
sales_df.loc[OUT045_filter].head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
8            FDH17       16.200          Regular         0.016687   
33           FDO23       17.850          Low Fat         0.000000   
46           NCP05       19.600          Low Fat         0.000000   
47           FDV49       10.000          Low Fat         0.025880   
56           NCX54        9.195          Low Fat         0.048157   

             Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
8         Frozen Foods   96.9726            OUT045                       2002   
33              Breads   93.1436            OUT045                       2002   
46  Health and Hygiene  153.3024            OUT045                       2002   
47              Canned  265.2226            OUT045                       2002   
56           Household  106.1622            OUT045                       2002   

   Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
8        Small               Tier 2  Supermarket Type1          1076.5986  
33       Small               Tier 2  Supermarket Type1          2174.5028  
46       Small               Tier 2  Supermarket Type1          2428.8384  
47       Small               Tier 2  Supermarket Type1          5815.0972  
56       Small               Tier 2  Supermarket Type1          2117.2440

##### Changing OUT010

In [251]:
# Change 'Outlet_Size' of all produce in OUT010 to 'Small'
OUT010_filter = sales_df['Outlet_Identifier'] == 'OUT010'
sales_df.loc[OUT010_filter, 'Outlet_Size'] = 'Small'
# Check that the size for OUT010 has changed
sales_df.loc[OUT010_filter].head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3            FDX07       19.200          Regular         0.000000   
28           FDE51        5.925          Regular         0.161467   
30           FDV38       19.250          Low Fat         0.170349   
45           FDM39        6.420               LF         0.089499   
65           FDC46       17.700               LF         0.195068   

                Item_Type  Item_MRP Outlet_Identifier  \
3   Fruits and Vegetables  182.0950            OUT010   
28                  Dairy   45.5086            OUT010   
30                  Dairy   55.7956            OUT010   
45                  Dairy  178.1002            OUT010   
65            Snack Foods  185.4266            OUT010   

    Outlet_Establishment_Year Outlet_Size Outlet_Location_Type    Outlet_Type  \
3                        1998       Small               Tier 3  Grocery Store   
28                       1998       Small               Tier 3  Grocery Store   
30                       1998       Small               Tier 3  Grocery Store   
45                       1998       Small               Tier 3  Grocery Store   
65                       1998       Small               Tier 3  Grocery Store   

    Item_Outlet_Sales  
3            732.3800  
28           178.4344  
30           163.7868  
45           358.2004  
65           184.4266

In [252]:
# Check new sum of missing values in Outlet_Size column
sales_df['Outlet_Size'].isna().sum()

0

### 6. Confirm that there are no missing values after addressing them

In [253]:
# Check the sum of the missing values in each column to check that all have been addressed
sales_df.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

### 7. Find and fix any inconsistent categories of data.

> Columns that should have a specific number of options are:
- Item_Fat_Content
- Item_Type
- Outlet_Size
- Outlet_Location_Type
- Outlet_Type
>
> I will check theses columns for inconsistent categories

In [254]:
# Check name in the Item_Fat_Content column
sales_df.value_counts('Item_Fat_Content')

Item_Fat_Content
Low Fat    5088
Regular    2886
LF          316
reg         117
low fat     112
dtype: int64

> The only two options in this column should be 'Low Fat' and 'Regular', I will have to fix the other naming conventions to fit with these two options

In [255]:
# Check names in the Item_Type column
sales_df.value_counts('Item_Type')

Item_Type
Fruits and Vegetables    1232
Snack Foods              1199
Household                 910
Frozen Foods              855
Dairy                     681
Canned                    649
Baking Goods              647
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
dtype: int64

> No names look inconsistent in this column

In [256]:
# Check names in the Outlet_Size column
sales_df.value_counts('Outlet_Size')

Outlet_Size
Small     4797
Medium    2790
High       932
dtype: int64

> No names look inconsistent in this column

In [257]:
# Check names in the Outlet_Location_Type column
sales_df.value_counts('Outlet_Location_Type')

Outlet_Location_Type
Tier 3    3347
Tier 2    2785
Tier 1    2387
dtype: int64

> No names look inconsistent in this column

In [258]:
# Check names in the Outlet_Type column
sales_df.value_counts('Outlet_Type')

Outlet_Type
Supermarket Type1    5577
Grocery Store        1082
Supermarket Type3     932
Supermarket Type2     928
dtype: int64

> No names look inconsistent in this column

#### Fix data in Item_Fat_Content column

In [259]:
# Change name 'LF' to 'Low Fat' in Item_Fat_Content column
sales_df['Item_Fat_Content'] = sales_df['Item_Fat_Content'].replace('LF', 'Low Fat')

In [260]:
# Change name 'reg' to 'Regular' in Item_Fat_Content column
sales_df['Item_Fat_Content'] = sales_df['Item_Fat_Content'].replace('reg', 'Regular')

In [261]:
# Change name 'low fat' to 'Low Fat' in Item_Fat_Content column
sales_df['Item_Fat_Content'] = sales_df['Item_Fat_Content'].replace('low fat', 'Low Fat')

In [262]:
# Check names in the Item_Fat_Content column have been changed
sales_df.value_counts('Item_Fat_Content')

Item_Fat_Content
Low Fat    5516
Regular    3003
dtype: int64

### 8. For any numerical columns, obtain the summary statistics of each (min, max, mean)

In [263]:
# Use .describe() to obtain summary statistics of each numerical column
sales_df.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  8519.000000      8519.000000  8519.000000                8519.000000   
mean     12.875420         0.066112   141.010019                1997.837892   
std       4.646098         0.051586    62.283594                   8.369105   
min       4.555000         0.000000    31.290000                1985.000000   
25%       8.785000         0.026983    93.844900                1987.000000   
50%      12.650000         0.053925   143.047000                1999.000000   
75%      16.850000         0.094558   185.676600                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8519.000000  
mean         2181.188779  
std          1706.511093  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3100.630600  
max         13086.964800

## Exploratory Visuals

## Explanatory Visuals